src/Microsoft.ApiDesignGuidelines.Analyzers/Core/TypeNamesShouldNotMatchNamespaces.cs

```diff
@@ -25,24 +29,236 @@ public abstract class TypeNamesShouldNotMatchNamespacesAnalyzer : DiagnosticAnal
         internal static DiagnosticDescriptor SystemRule = new DiagnosticDescriptor(RuleId,
                                                                              s_localizableTitle,
                                                                              s_localizableMessageSystem,
                                                                              DiagnosticCategory.Naming,
                                                                              DiagnosticSeverity.Warning,
-                                                                             isEnabledByDefault: false,
+                                                                             isEnabledByDefault: true,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182257.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule, SystemRule);
 
+        private static object s_obj = new object();
+        private static ImmutableDictionary<string, string> s_wellKnownSystemNamespaceTable;
+
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(
+                compilationStartAnalysisContext =>
+                {
+                    var namedTypesInCompilation = new ConcurrentBag<INamedTypeSymbol>();
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        SymbolAnalysisContext =>
+                        {
+                            var namedType = (INamedTypeSymbol)SymbolAnalysisContext.Symbol;
+                            namedTypesInCompilation.Add(namedType);
+                        }, SymbolKind.NamedType);
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        compilationAnalysisContext =>
+                        {
+                            var namespaceNamesInCompilation = new ConcurrentBag<string>();
+                            var compilation = compilationAnalysisContext.Compilation;
+                            AddNamespacesFromCompilation(namespaceNamesInCompilation, compilation.GlobalNamespace);
+
+                            // The entries are added in the sorted order so that the diagnostic message is always deterministic
+                            var namespaceTableForCompilation = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```

Comment:
It took me a while to figure out what you were doing here, so I suggest a variable name change and an expanded comment on Line 72.

Consider renaming `namespaceTableForCompilation` to `namespaceComponentToNamespaceNameDictionary` as the variable name, and perhaps this comment in place of the one on Line 72:

> We construct a dictionary whose keys are all the components of all the namespace names in the compilation, and whose values are the namespace names of which the components are a part. For example, if the compilation includes namespaces A.B and C.D, the dictionary will map "A" to "A.B", "B" to "A.B", "C" to "C.D", and "D" to "C.D". When the analyzer encounters a type name that appears in a dictionary, it will emit a diagnostic, for instance, "Type name "D" conflicts with namespace name "C.D"".

> A component can occur in more than one namespace (for example, you might have namespaces "A.B" and "C.A".) In that case, we have to choose one namespace to report the diagnostic on. We want to make sure that this is deterministic (we don't won't to complain about "A.B" in one compilation, and about "C.A" in the next). By calling `ToImmutableSortedSet` on the list of namespace names in the compilation, we ensure that we'll always construct the dictionary with the same set of keys.